### ISSUE TITLE
Update MSC3912 implementation: Redaction of related events

### ISSUE THREAD

**Issue by giomfo** [matrix-org/synapse#15687]
**Description:**

Since MSC3912 has been implemented by https://github.com/matrix-org/synapse/pull/14260, we introduced [2 changes](https://github.com/matrix-org/matrix-spec-proposals/pull/3912/commits/55060fc3b4634215d417b4cf280e8cda1bb64613) into the MSC for which we need to update this implementation :
- the new property `with_relations` has been renamed `with_rel_types` (to prevent us from introducing a new term ("relations") to refer to event relationships). Only the stable name is changed, we keep the existing unstable one.
- we introduced a catch-all `"*"` value, which if found in the list means "any relation type"

**Comment by giomfo** [comment]
> There is one stuff missing:

>> This implementation is lacking a way to keep watch for incoming events that match the redaction criteria after the redaction request finishes, that'll come in a separate PR.

> It's a bit less simple since we need to store the redaction action somewhere in DB. I am planning to have a look a bit later.

@wrjlewis Hi, about this opened request, can you confirm that we will not implement in synapse the last remaining point? I would like to unblock the MSC3912 review by resolving the last opened conversation [here](https://github.com/matrix-org/matrix-spec-proposals/pull/3912/files#r1211499254). Feel free to contact me in DM if you need more context.

**Comment by erikjohnston** [comment]
I think it would take a week or two of effort to implement. Having had a brief glance I think there are a few plausible ways of doing this:

- Check if an event with a relation is redacted via these new rules when we *read* from the DB; or
- Check when we persist a new event whether it matches a redaction of one of these rules.

Not sure if we'ed need to add a special table for recording these new sort of redactions or not.

I think the question is now: do we care enough about that MSC to spend the time implementing this now or at some point later?
